### PR TITLE
[WIP][DO NOT REVIEW] bench: add router CPU fast-path baseline harness

### PR DESCRIPTION
**Pull request by jiannan-17**
## Purpose

Add benchmarks to measure router overhead before the tokenizer/L0 request-path integration in #244, following #270. No runtime behavior changes.

- Criterion benchmarks for routing-key extraction, worker selection, and the tree-key formats discussed in #237.
- An opt-in harness that runs the router against mock workers and reports client latency, throughput, router CPU time, and peak RSS. A direct-to-worker run provides a reference.

Both use a shared, deterministic corpus. Run instructions and measurement details are in `docs/benchmarks/router_overhead.md`. These benchmarks measure router overhead; routing benefits need real-worker measurements.

## Test Plan

```text
cargo test --test bench_corpus_test
cargo bench --bench routing_input -- --test
cargo bench --bench routing_input
cargo test --release --test router_overhead_bench -- --ignored --nocapture
cargo +1.95.0 clippy --all-targets --all-features -- -D warnings
```

## Test Result

Validation recorded during implementation:

- Corpus tests: 7 passed.
- Criterion smoke run: all groups ran.
- Clippy (Rust 1.95.0) and codespell: clean.

Benchmark measurements are pending. The tree-key format comparison also needs review before this is ready.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results

</details>

### CODE REVIEW

File: scripts/run_benchmarks.py [hunk 1]
```diff
@@ -220,16 +256,45 @@ def main():
     # Build in release mode
     runner.build_release()
 
+    only_request_processing = [
+        name
+        for name, used in (
+            ("--quick", args.quick),
+            ("--save-baseline", args.save_baseline),
+            ("--compare-baseline", args.compare_baseline),
```
[the remark this answers is not in the document]
Comment (jiannan-17) [reply scripts/run_benchmarks.py]: Fixed in 83ac4ab. --save-baseline already works with --bench routing_input. it’s only ignored by --router-overhead. I updated the note to list only the options ignored by the selected run.

File: tests/router_overhead_bench.rs [hunk 2]
```diff
@@ -0,0 +1,968 @@
...
+    },
+    Scenario {
+        name: "completions_rendezvous",
+        policy: Some("rendezvous_hash"),
+        route: Route::Completions,
+    },
+    Scenario {
+        name: "chat_off",
+        policy: Some("cache_aware"),
+        route: Route::Chat,
```
[the remark this answers is not in the document]
Comment (jiannan-17) [reply tests/router_overhead_bench.rs]: Good catch, fixed in 83ac4ab. I added direct_chat and renamed direct_mock to direct_completions, so each route now uses its own baseline. If that baseline wasn’t run, the delta is left blank. A new test checks that every routed scenario has a matching direct baseline.

Measured both ways on the same run (Linux, hot64, median of 3):

| cell | old (completions floor) | new (chat floor) |
|---|---:|---:|
| chat_off 200 B, c=1 | 79.1 us | 78.7 us |
| chat_off 16 KiB, c=1 | 94.0 us | 93.4 us |
| chat_off 2 KiB, c=64 | 2322.7 us | 2259.3 us |
| chat_off 16 KiB, c=64 | 2917.2 us | 2822.3 us |

The direct baselines differ by about 1 µs at concurrency 1 and 95 µs at concurrency 64. Using the completions baseline overstated chat overhead by roughly 3–4% under load. The PR description only reports /v1/completions results, so those numbers are unaffected.

File: tests/router_overhead_bench.rs [hunk 3]
```diff
@@ -0,0 +1,968 @@
...
+                .map(|s| s.parse().expect("size"))
+                .collect(),
+            corpora: env_list("VLLM_ROUTER_BENCH_CORPORA", "hot64,cold")
+                .iter()
+                .map(|c| CorpusKind::parse(c).unwrap_or_else(|| panic!("unknown corpus {c}")))
+                .collect(),
+            concurrency: env_list("VLLM_ROUTER_BENCH_CONCURRENCY", "1,64")
+                .iter()
+                .map(|c| c.parse().expect("concurrency"))
+                .collect(),
```
[the remark this answers is not in the document]
Comment (jiannan-17) [reply tests/router_overhead_bench.rs]: Fixed in 83ac4ab: `Config::from_env` now asserts every value in `VLLM_ROUTER_BENCH_CONCURRENCY` is at least 1, so `0` fails at startup instead of reporting an empty run.

File: benches/routing_input.rs [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+//! derive the routing key from a typed request and to select a worker,
+//! before any tokenizer runs in the request path.
+//!
+//! Groups:
+//! - `routing_key/extract_text_for_routing/{completion,chat}/{size}`:
+//!   `GenerationRequest::extract_text_for_routing` as called once per request
+//!   in `Router::route_typed_request`, including dropping the returned
+//!   `String` as the router does. (`iter_with_large_drop` would keep a whole
+//!   sample's worth of 16 KiB strings alive and needs tens of GB.)
+//! - `policy/{cache_aware,rendezvous_hash}/{size}`: `select_worker` over four
```
Comment (jiannan-17) [inline benches/routing_input.rs]: add specific use cases edge cases to cover different corner cases

add long input such as 128k long, 1mb size

Comment (jiannan-17) [reply benches/routing_input.rs]: updated with more specific test cases.

## Test Results

All 20 tests passed on Linux and again locally after cleanup: five routing tests, 14 corpus tests and one harness check. Eight deliberate routing mutations were all caught. Corpus digests matched `83ac4ab` for the first 200 prompts of each existing corpus at each default size.

Linux, 16 vCPU, Rust 1.95.0, release build; four mock workers.

| Run | Repeats per configuration | Warmup + measurement | Concurrency | Measurements |
|------|---:|---|---|---:|
| Baseline, 200 B–16 KiB | 3 | 4 s + 10 s | 1, 64 | 360 |
| Pinned control, 2/16 KiB | 3 | 4 s + 10 s | 64 | 60 |
| Long text and token IDs | 1 | 3 s + 5 s | 1, 16 | 96 |

All 516 measurements completed without errors. Criterion covered 57 cases: one initial suite plus parsing reruns, with 1 s warmup, a 5 s measurement target and 50 samples (10 for inputs ≥512 KiB).

* Long ASCII hot-set prompts (concurrency 1): Cache-aware had 42–45% lower router overhead than rendezvous in one run.
* Small cold or shared-prefix prompts (200 B–2 KiB): Rendezvous had lower router overhead than cache-aware.
* CJK prompts: Router overhead was similar for both policies. Most of the time went into parsing ([#310](https://github.com/vllm-project/router/issues/310)).
* 16 KiB cold prompts (concurrency 64): Cache-aware had lower router overhead but used more memory. Median peak memory across three runs was 2,123 MiB for cache-aware and 77 MiB for rendezvous.

These compare existing policies; they are not performance gains from this PR. Mock workers measure router overhead without model execution or KV-cache benefits.

[Full results and JSON data](https://gist.github.com/jiannan-17/ee7c4332d6ace6e86b538f2c4b99bf47/62315a77f777fa949020147f3853d67784f2eec6#file-00-report-md) include per-case latency, throughput, memory, CPU, worker distribution, source versions and commands.
